### PR TITLE
4Clojure & AOC solutions

### PR DESCRIPTION
**Pull request by olttwa**

### CODE REVIEW

File: src/main.clj [hunk 1]
```diff
@@ -1,2 +1,24 @@
 ; https://4clojure.oxal.org/#/problem/1
 (= true true)
+
+; solved problems 2-18 in the web browser.
+
+; https://4clojure.oxal.org/#/problem/25
+(= (filter (fn [x] (= 1 (mod x 2))) #{1 2 3 4 5}) '(1 3 5))
```
Comment (kitallis) [inline src/main.clj]: This is good, but better yet, [odd?](https://clojuredocs.org/clojure.core/odd_q).

File: src/main.clj [hunk 2]
```diff
@@ -1,2 +1,24 @@
...
+
+; https://4clojure.oxal.org/#/problem/26
+((fn fibonacci
+   ([n]
+    (case n
+      0 ()
+      1 '(1)
+      2 '(1 1)
+      (fibonacci (- n 2) '(1 1))))
```
Comment (kitallis) [inline src/main.clj]: The way you have written it is non-lazy, which is perfectly fine.

In the real world, you'll seldom work with infinite streams of data, as it so happens that `fibonacci` is one of those examples that has an infinite stream solution / problem space. So lazy-seqs can work quite well here.

##### a little bit of diversion here
------


###### Most sequence functions in Clojure are lazy

i.e `map`, `filter`, `for` etc.

###### `PersistentList` is separate from a `LazySeq`

This is a `PersistentList`

```
'(1 2)
(class '(1 2))
``` 

This returns a `LazySeq`

```
(take 3 (range))
(class (take 3 (range)))
```

They both "look like" lists. Because, well, they are fundamentally. You can't get infinite / lazy sequences with indexed data-structures like vectors; you need something like a linked-list.

A persistent list is just something that is a realized list, i.e. you can do `count` in `O(1)`. A `LazySeq`, is just something that knows how to traverse the next item in the list, but hasn't yet.

```
(cons 3 '(1 2))
(class (cons 3 '(1 2)))
```

Will always return `Cons` object.

In other words, `conj` knows what you passed to it and returns you an appropriate collection back, but `cons` just needs anything that is an `ISeq` and always returns a `Cons` back.

`Cons` is what `LazySeqs` are made of.

All of the above, `PersistentVector`, `PersistentList`, `LazySeq` they all share a common interface under the sequence abstraction and hence work well with each other, generally.

If this seems confusing and/or overwhelming, I can explain this in-person with examples. Also, this is a good read on the topic: https://clojure.org/reference/sequences

------


There's potentially a neat lazy solution to this problem that purely relies on `cons` and lists.

Check out `lazy-seq` and `iterate` and `take`.

File: src/main.clj [hunk 3]
```diff
@@ -1,2 +1,24 @@
...
+   ([n s]
+    (if (< n 1)
+      s
+      (fibonacci (- n 1) (concat s [(reduce + (take-last 2 s))]))))
```
Comment (kitallis) [inline src/main.clj]: Note that, since the operation is "associative" in a way. `apply` will work equally well in place of a `reduce`.

**Review by kitallis — COMMENTED** [review]
Looking good so far. Left some thoughts. 🚀
I can edit?

File: src/main.clj (hunk 1 above)
Comment (olttwa) [reply src/main.clj]: modified

File: src/main.clj (hunk 2 above)
Comment (olttwa) [reply src/main.clj]: I hadn't been exposed to lazy-seq until this problem. Have used it in problem #114.

File: src/main.clj (hunk 3 above)
Comment (olttwa) [reply src/main.clj]: I came across `apply` after this function. Have began using this from now on 😅

File: src/main.clj [hunk 4]
```diff
@@ -1,2 +1,314 @@
...
+; Double, Halve or Add 2
+; Unsolved
+
+; https://4clojure.oxal.org/#/problem/73
+; This solution isn't extensible :( 😭
+((fn tic-tac-toe
+   ([board] (tic-tac-toe 0 board))
+   ([c board]
+    ; Iterate over all rows to find a winner.
+    (let [winner (reduce (fn [_ r]
```
Comment (kitallis) [inline src/main.clj:158]: I think this `reduce` can definitely be changed to a `filter`. It will short-circuit.

File: src/main.clj [hunk 5]
```diff
@@ -1,2 +1,314 @@
...
+                            (map-indexed (fn [i e] (get e (- 2 i))) board)])
+          2 nil)))))
+ [[:x :e :e]
+  [:o :x :e]
+  [:o :e :x]])
+
+; https://adventofcode.com/2021/day/1
+((defn aoc1-part1
+   [file]
+   (with-open [rdr (clojure.java.io/reader file)]
```
Comment (kitallis) [inline src/main.clj:180]: Would recommend simply the reading the file up, since it's finite. For all AoC solutions.

See if you can make use of `->>` across `io/reader`, `line-seq` and `parseInt`.

File: src/main.clj [hunk 6]
```diff
@@ -1,2 +1,314 @@
...
+          (line-seq rdr))]
+       (* (:horizontal position) (:depth position)))))
+ "/Users/a/w/4clojure/aoc2-input.txt")
+
+; Mars Rover - https://github.com/priyaaank/MarsRover
+; QQQ: How to achieve extensibility in Clojure?
+; The way I see it, all inter-method calls/reduce are deeply coupled.
+; And if I were to look at this after a few months, I won't be able to make head/tail out of this :(
+(defn parse-input
+  [[x y d]]
```
Comment (jysandy) [inline src/main.clj:267]:
```suggestion
  [[x y direction]]
```
This name becomes much easier to understand if the name is spelled out. 
In general, prefer complete spellings over abbreviations unless absolutely necessary. With autocomplete features in modern IDEs, abbreviations shouldn't be necessary to save typing effort. Single letter variable names are almost always a no-no.

File: src/main.clj [hunk 7]
```diff
@@ -1,2 +1,314 @@
...
+ [[:x :e :e]
+  [:o :x :e]
+  [:o :e :x]])
+
+; https://adventofcode.com/2021/day/1
+((defn aoc1-part1
+   [file]
+   (with-open [rdr (clojure.java.io/reader file)]
+     (:count
+       (reduce
```
Comment (kitallis) [inline src/main.clj:182]: Instead of large reducers, try and break the problem down into smaller units. Think pipelines not procedures. Take some input, process it, pass it over, transform it, repeat.

This problem can be much simply solved by just comparing the last value of the current step against the first element of the previous step since the middle values are the same. Since the rolling window has repeating elements.

```
[1 3 5]
[3 5 7]
[5 7 9]
```

Comparing 1 with 7, 3 with 9 etc.

Here's my ruby solution for reference (written in a "functional" way):

```ruby
def aoc1(partition, step)
  File.read("input1")
      .lines
      .map(&:to_i)
      .each_cons(partition)
      .filter_map { |e| e[step] > e[0] }
      .size
end

aoc1(2,1) # part 1
aoc1(4,3) # part 2
```

File: src/main.clj [hunk 8]
```diff
@@ -1,2 +1,314 @@
...
+       (* (:horizontal position) (:depth position)))))
+ "/Users/a/w/4clojure/aoc2-input.txt")
+
+; Mars Rover - https://github.com/priyaaank/MarsRover
+; QQQ: How to achieve extensibility in Clojure?
+; The way I see it, all inter-method calls/reduce are deeply coupled.
+; And if I were to look at this after a few months, I won't be able to make head/tail out of this :(
+(defn parse-input
+  [[x y d]]
+  [(Integer/parseInt x) (Integer/parseInt y) (keyword d)])
```
Comment (jysandy) [inline src/main.clj:268]: Prefer maps over vectors for modelling associative data. A rover has an x-coordinate, a y-coordinate and a direction. It might look like this:
```
{:x 5
 :y 6
 :direction :N}
```
With the vector modelling, one has to carry around the tacit knowledge that the first item is the x-coordinate, the second item is the y-coordinate and the third item is the direction. It's easier to use a map and give them names directly.
When using a vector or a sequence, ask yourself if the order of elements should matter. If not, you may be better off using a map or a set.

**Review by kitallis — COMMENTED** [review]
First pass.

File: src/main.clj [hunk 9]
```diff
@@ -1,2 +1,314 @@
...
+          {:horizontal 0 :depth 0 :aim 0}
+          ; Feed sequence of readings to reduce.
+          (line-seq rdr))]
+       (* (:horizontal position) (:depth position)))))
+ "/Users/a/w/4clojure/aoc2-input.txt")
+
+; Mars Rover - https://github.com/priyaaank/MarsRover
+; QQQ: How to achieve extensibility in Clojure?
+; The way I see it, all inter-method calls/reduce are deeply coupled.
+; And if I were to look at this after a few months, I won't be able to make head/tail out of this :(
```
Comment (jysandy) [inline src/main.clj:265]: I agree that as of now, this code (and even the AoC solutions) aren't easy to understand.

This is mostly because the problems haven't been broken down well. After reading the Mars Rover problem statement, one might intuitively come up with the following rough solution:

1. The input is a series of rover positions and associated commands.
2. We first find out how to compute a rover's final position given an initial position and its commands.
3. Now that we know how to do this, we apply this procedure to the input to get a series of final positions.

However, this structure is not reflected in the solution here. The solution is written as one big `reduce`, and most of the domain logic is mixed up in the body. The same applies to the AoC solutions.

Let's discuss more over a call.

File: src/main.clj [hunk 10]
```diff
@@ -1,2 +1,314 @@
...
+(defn parse-input
```
Comment (jysandy) [inline src/main.clj:266]:
```suggestion
(defn parse-rover-state
```
This input is quite specific (it's the state of a rover) but the name is quite generic (what input?).

File: src/main.clj (hunk 9 above)
Comment (jysandy) [reply src/main.clj:265]: Summary of our discussion:

1. Break down the problem through domain modelling. The code should reflect the intuitive solution to the problem in our head.
2. Define data models and prefer maps over vectors for associative data. One doesn't require types to define a data model.
3. Loops, tail recursion and `reduce` are very low-level constructs. Prefer higher level functions like `map`, `filter`, `partition` or other higher-level domain functions.
